src/codegen/operator/csv_scan_translator.cpp

```diff
@@ -0,0 +1,307 @@
+    auto *data_ptr = codegen->CreateLoad(codegen->CreateConstInBoundsGEP2_32(
+        CSVScannerColumnProxy::GetType(codegen), Columns(), ColumnIndex(), 1));
+    auto *data_len = codegen->CreateLoad(codegen->CreateConstInBoundsGEP2_32(
+        CSVScannerColumnProxy::GetType(codegen), Columns(), ColumnIndex(), 2));
+
+    // If the valid isn't NULLable, avoid the null check here
+    if (!IsNullable()) {
+      return LoadValueIgnoreNull(codegen, type, data_ptr, data_len);
+    }
+
+    // If the value isn't actually null, try to parse it
+    codegen::Value valid_val, null_val;
+    lang::If is_null{codegen,
+                     codegen->CreateNot(IsNull(codegen, data_ptr, data_len))};
+    {
+      // Load valid
+      valid_val = LoadValueIgnoreNull(codegen, type, data_ptr, data_len);
+    }
+    is_null.ElseBlock();
+    {
+      // Default null
+      null_val = SqlType().GetNullValue(codegen);
+    }
+    is_null.EndIf();
+
+    // Return
+    return is_null.BuildPHI(valid_val, null_val);
+  }
+
+ private:
+  // Information about the attribute
+  const planner::AttributeInfo *ai_;
+
+  // A pointer to the array of columns
+  llvm::Value *csv_columns_;
+
+  // The NULL string configured for the CSV scan
+  const std::string null_str_;
+
+  // The runtime NULL string (a constant in LLVM)
+  llvm::Value *runtime_null_;
+};
+
+}  // namespace
+
+// We define the callback/consumer function for CSV parsing here
+void CSVScanTranslator::DefineAuxiliaryFunctions() {
+  CodeGen &codegen = GetCodeGen();
+  CompilationContext &cc = GetCompilationContext();
+
+  auto &scan = GetPlanAs<planner::CSVScanPlan>();
+
+  // Define consumer function here
+  std::vector<FunctionDeclaration::ArgumentInfo> arg_types = {
+      {"queryState", cc.GetQueryState().GetType()->getPointerTo()}};
+  FunctionDeclaration decl{codegen.GetCodeContext(), "consumer",
+                           FunctionDeclaration::Visibility::Internal,
+                           codegen.VoidType(), arg_types};
+  FunctionBuilder scan_consumer{codegen.GetCodeContext(), decl};
+  {
+    ConsumerContext ctx{cc, GetPipeline()};
+
+    Vector v{nullptr, 1, nullptr};
+    RowBatch one{GetCompilationContext(), codegen.Const32(0),
+                 codegen.Const32(1), v, false};
+
+    // Load the pointer to the columns view
+    llvm::Value *cols = codegen->CreateLoad(codegen->CreateConstInBoundsGEP2_32(
+        CSVScannerProxy::GetType(codegen), LoadStatePtr(scanner_id_), 0, 1));
+
+    llvm::Value *null_str = codegen.ConstString(scan.GetNullString(), "null");
+
+    // Add accessors for all columns into the row batch
+    std::vector<CSVColumnAccess> column_accessors;
+    for (uint32_t i = 0; i < output_attributes_.size(); i++) {
+      column_accessors.emplace_back(output_attributes_[i], cols,
+                                    scan.GetNullString(), null_str);
+    }
+    for (uint32_t i = 0; i < output_attributes_.size(); i++) {
```

Comment:
It needs to be two loops because we build up a vector in the first loop and insert pointers to vector elements in the second loop. If it's in the same loop, the pointers may be invalid due to resizing.

We **could** reserve space beforehand and use one loop. We use this pattern in other places.